chore: archiver validates L1->L2 messages

LeilaWang opened the pull request:
- Move the validation of the `inHash` in the checkpoint header from `world_state_synchroniser` to `archiver`.
- Add `checkpointNumber` and `indexWithinCheckpoint` to `L2BlockNew` to help identify the first block in a checkpoint when processing L1->L2 messages and encoded blob data.
- Everything L1->L2 message related is now associated with checkpoint.
  - Except for the user facing api `aztecNode.getL1ToL2MessageBlock()`, which is still using block number to fetch L1->L2 messages.

LeilaWang commented on yarn-project/stdlib/src/block/l2_block_new.ts:
```diff
@@ -22,6 +22,10 @@ export class L2BlockNew {
+    /** Number of the checkpoint that the block belongs to. */
+    public checkpointNumber: CheckpointNumber,
+    /** Index of the block within the checkpoint. */
+    public indexWithinCheckpoint: number,
```
Thanks @spalladino for the suggestion! Everything is much cleaner now 💜

spalladino commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -878,7 +878,21 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
+            },
+          );
+
+          continue;
```
IIUC we used to throw if there was an error converting the checkpoint object, which halted the archiver sync, whereas we are now continuing, right? Why the change? Not saying we shouldn't, just wondering about the rationale for it since it's a big change. Also, do we know what could trigger an error in the conversion here (aside from a bug)?

LeilaWang replied:
I changed it this way to match what we do when the attestations are invalid. And when processing the logs from  L1 in `data_retrieval.ts`, we also simply "skip" the checkpoint if the archive roots do not match (although I'm not sure if that's possible).

If the proposer is honest, it's most likely a bug that triggers an error from the conversion. Same for the inHash below.

Let me know if you think throwing an error is preferable and I'll change it!

spalladino replied:
We purposefully skip the block if attestations are invalid since we know that the next proposer will "invalidate" this block, removing it from the chain. And we skip the block if the archive root does not match the one in rollup storage because it means we're trying to sync a block that has been pruned. In both situations the block we're skipping is not part of the chain.

That's why I'm worried about skipping in other situations. If we have a bug and there's a _valid_ block we can't sync, then we should not skip it, since the following block would fail due to the gap in the chain. So I think I'd go back to failing loudly. WDYT?

spalladino replied:
By the way, I'll approve so you can merge this one if you want to change the error handling in another PR. Whatever you prefer!

LeilaWang replied:
Makes perfect sense! I'll change the encoding and in hash to throw an error instead. Thanks for explaining it 🙌

spalladino replied:
Since you're at it, can you add a comment explaining the above (or just copy pasting it)? I realized that logic is not documented anywhere, my fault.

spalladino commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -916,6 +930,22 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
+            publishedInHash: published.checkpoint.header.contentCommitment.inHash,
+          });
+
+          continue;
```
Same here

spalladino reviewed (commented):
Thanks @LeilaWang! Changes look good, I just have a question about error handling. Are we sure about continuing the archiver sync loop on an unexpected error? Or should we just abort? Can a single proposer trigger an error that halts archivers, without help from the committee?

LeilaWang committed:
chore: archiver validates L1->L2 messages

- Move the validation of the `inHash` in the checkpoint header from `world_state_synchroniser` to `archiver`.
- Add `checkpointNumber` and `indexWithinCheckpoint` to `L2BlockNew` to help identify the first block in a checkpoint when processing L1->L2 messages and encoded blob data.
- Everything L1->L2 message related is now associated with checkpoint.
  - Except for the user facing api `aztecNode.getL1ToL2MessageBlock()`, which is still using block number to fetch L1->L2 messages.

Co-authored-by: Leila Wang <leizciw@gmail.com>
